enterprise_access/apps/api/v1/exceptions.py

```diff
@@ -0,0 +1,143 @@
+        """
+        super().__init__(code=code, detail=detail)
+
+        self.reason = self.default_reason
+        self.user_message = self.default_user_message
+        self.metadata = {}
+
+        if policy and subsidy_api_error:
+            try:
+                self._build_subsidy_api_error_payload(policy, subsidy_api_error)
+            except Exception:  # pylint: disable=broad-except
+                self.metadata = {
+                    'subsidy_error_detail_raw': str(subsidy_api_error.error_response.content),
+                }
+
+        self.detail = {
+            'code': self.code,
+            'detail': {
+                'reason': self.reason,
+                'user_message': self.user_message,
+                'metadata': self.metadata,
+            }
+        }
+
+    def _build_subsidy_api_error_payload(self, policy, subsidy_api_error):
+        """
+        Helper to build error response payload on Subsidy API errors.
+        """
+        subsidy_error_detail = subsidy_api_error.error_payload().get('detail')
+        subsidy_error_code = subsidy_api_error.error_payload().get('code')
+
+        self.metadata = {
+            'enterprise_administrators': LmsApiClient().get_enterprise_customer_data(
+                policy.enterprise_customer_uuid
+            ).get('admin_users')
+        }
+
+        # We currently only have enough data to say more specific things
+        # about fulfillment errors during subsidy API redemption.
+        if subsidy_error_code == constants.SubsidyRedemptionErrorCodes.FULFILLMENT_ERROR:
+            self._set_subsidy_fulfillment_error_reason(subsidy_error_detail)
+
+    def _set_subsidy_fulfillment_error_reason(self, subsidy_error_detail):
+        """
+        Helper to set the reason, user_message, and metadata
+        for the given subsidy_error_detail.
+        """
+        self.metadata['subsidy_error_detail'] = subsidy_error_detail
+        self.reason = constants.SubsidyFulfillmentErrorReasons.DEFAULT_REASON
+
+        if subsidy_error_detail:
+            message_string = self._get_subsidy_fulfillment_error_message(subsidy_error_detail)
+            if cause_of_message := constants.SubsidyFulfillmentErrorReasons.get_cause_from_error_message(
+                message_string
+            ):
+                self.reason = cause_of_message
+                # pylint: disable=attribute-defined-outside-init
+                self.code = constants.SubsidyRedemptionErrorCodes.FULFILLMENT_ERROR
+
+        self.user_message = constants.SubsidyFulfillmentErrorReasons.USER_MESSAGES_BY_REASON.get(self.reason)
+
+    def _get_subsidy_fulfillment_error_message(self, subsidy_error_detail):
+        """
+        ``subsidy_error_detail`` is either a string describing an error message,
+        a dict with a "message" key describing an error message, or a list of such
+        dicts.  This helper method widdles any of those things down into a single
+        error message string.
```

Comment:
Is this just the result of the subsidy API being indecisive about error formatting and being written by >1 person?

Reply:
Oof yes, can we fix the offending call?

Reply:
I wrote to be defensive about the types of messages that can exist from a DRF APIException: https://github.com/encode/django-rest-framework/blob/master/rest_framework/exceptions.py